Clustered buffer underflow seen when under pressure

o.e.c.i.r.LoggingRobustResilienceStrategy - Ehcache key 45099 in possible inconsistent state due to 
org.ehcache.core.spi.store.StoreAccessException: java.nio.BufferUnderflowException
	at org.ehcache.core.exceptions.StorePassThroughException.handleRuntimeException(StorePassThroughException.java:72) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.clustered.client.internal.store.ClusteredStore.silentPut(ClusteredStore.java:189) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.clustered.client.internal.store.ClusteredStore.put(ClusteredStore.java:160) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.impl.internal.store.tiering.TieredStore.put(TieredStore.java:149) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.core.Ehcache.put(Ehcache.java:196) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at com.terracotta.qa.ehcache.validation.ValidatingPutOperation.exec(ValidatingPutOperation.java:56) [classes/:na]
	at io.rainfall.execution.RunsDuring$1.call(RunsDuring.java:79) [rainfall-core-1.1.3-20160624.105609-14.jar:na]
	at io.rainfall.execution.RunsDuring$1.call(RunsDuring.java:71) [rainfall-core-1.1.3-20160624.105609-14.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_66]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_66]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_66]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_66]
Caused by: java.nio.BufferUnderflowException: null
	at java.nio.Buffer.nextGetIndex(Buffer.java:500) ~[na:1.8.0_66]
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:135) ~[na:1.8.0_66]
	at org.ehcache.clustered.client.internal.store.operations.codecs.OperationsCodec.decode(OperationsCodec.java:41) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.clustered.client.internal.store.operations.ChainResolver.resolve(ChainResolver.java:65) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	at org.ehcache.clustered.client.internal.store.ClusteredStore.silentPut(ClusteredStore.java:182) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT dbaf9901e0e2786eef9e107b174c3f9a869650cf]
	... 10 common frames omitted

--- a/clustered/common/src/main/java/org/ehcache/clustered/common/internal/store/ServerStore.java
+++ b/clustered/common/src/main/java/org/ehcache/clustered/common/internal/store/ServerStore.java
@@ -49,6 +49,7 @@ public interface ServerStore {
   /**
    * Appends the provided binary to Chain associated with key atomically.
    * While appending, the payLoad is stored in {@link Element}.
+   * Note that the {@code payLoad}'s position and limit are left untouched.
    *
    * @param key to which the payLoad has to be appended
    * @param payLoad to be appended
@@ -66,6 +67,7 @@ public interface ServerStore {
    *    append(key, payLoad);
    *    return prevChain;
    *  }
+   * Note that the {@code payLoad}'s position and limit are left untouched.
    *
    * @param key to which the payLoad has to be appended
    * @param payLoad to be appended
@@ -103,6 +105,7 @@ public interface ServerStore {
    *    hash -> |payLoadC"| - |payLoadD|
    *
    * then replaceAtHead(hash, |payLoadA| - |payLoadB| - |payLoadC|, |payLoadC'|) will be ignored.
+   * Note that the payload's position and limit of all elements of both chains are left untouched.
    *
    * @param key for which Chain has to be replaced
    * @param expect of the original Chain which this operation intends to replaceAtHead

--- a/clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainStorageEngine.java
+++ b/clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainStorageEngine.java
@@ -447,7 +447,7 @@ class OffHeapChainStorageEngine<K> implements StorageEngine<K, InternalChain> {
   private long writeElement(long address, ByteBuffer element) {
     storage.writeLong(address + ELEMENT_HEADER_SEQUENCE_OFFSET, nextSequenceNumber++);
     storage.writeInt(address + ELEMENT_HEADER_LENGTH_OFFSET, element.remaining());
-    storage.writeBuffer(address + ELEMENT_HEADER_SIZE, element);
+    storage.writeBuffer(address + ELEMENT_HEADER_SIZE, element.duplicate());
     return address;
   }
 

--- a/clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java
+++ b/clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java
@@ -154,7 +154,7 @@ public class OffHeapServerStore implements ServerStore {
     }
   }
 
-  private OffHeapChainMap<Long> segmentFor(long key) {
+  OffHeapChainMap<Long> segmentFor(long key) {
     return segments.get(Math.abs((int) (key % segments.size())));
   }
 
@@ -170,7 +170,7 @@ public class OffHeapServerStore implements ServerStore {
     }
   }
 
-  private boolean handleOversizeMappingException(long hash) {
+  boolean handleOversizeMappingException(long hash) {
     boolean evicted = false;
 
     OffHeapChainMap<Long> target = segmentFor(hash);

--- a/clustered/server/src/test/java/org/ehcache/clustered/server/offheap/OffHeapServerStoreTest.java
+++ b/clustered/server/src/test/java/org/ehcache/clustered/server/offheap/OffHeapServerStoreTest.java
@@ -16,14 +16,29 @@
 package org.ehcache.clustered.server.offheap;
 
 import java.nio.ByteBuffer;
+
 import org.ehcache.clustered.common.internal.store.Chain;
 import org.ehcache.clustered.common.internal.store.Element;
 import org.ehcache.clustered.server.store.ChainBuilder;
 import org.ehcache.clustered.server.store.ElementBuilder;
 import org.ehcache.clustered.common.internal.store.ServerStore;
 import org.ehcache.clustered.server.store.ServerStoreTest;
+import org.junit.Test;
+import org.mockito.Matchers;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
 import org.terracotta.offheapstore.buffersource.OffHeapBufferSource;
+import org.terracotta.offheapstore.exceptions.OversizeMappingException;
 import org.terracotta.offheapstore.paging.UnlimitedPageSource;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
 
 public class OffHeapServerStoreTest extends ServerStoreTest {
 
@@ -59,6 +74,91 @@ public class OffHeapServerStoreTest extends ServerStoreTest {
         };
       }
     };
+  }
+
+  @Test
+  public void test_append_doesNotConsumeBuffer_evenWhenOversizeMappingException() throws Exception {
+    OffHeapServerStore store = (OffHeapServerStore) spy(newStore());
+    final OffHeapChainMap offHeapChainMap = mock(OffHeapChainMap.class);
+    doThrow(OversizeMappingException.class).when(offHeapChainMap).append(Matchers.any(), any(ByteBuffer.class));
+
+    when(store.segmentFor(anyLong())).then(new Answer<Object>() {
+      int invocations = 0;
+      @Override
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        if (invocations++ < 10) {
+          return offHeapChainMap;
+        } else {
+          return invocation.callRealMethod();
+        }
+      }
+    });
+    when(store.handleOversizeMappingException(anyLong())).thenReturn(true);
+
+
+    ByteBuffer payload = createPayload(1L);
+
+    store.append(1L, payload);
+    assertThat(payload.remaining(), is(8));
+  }
+
+  @Test
+  public void test_getAndAppend_doesNotConsumeBuffer_evenWhenOversizeMappingException() throws Exception {
+    OffHeapServerStore store = (OffHeapServerStore) spy(newStore());
+    final OffHeapChainMap offHeapChainMap = mock(OffHeapChainMap.class);
+    doThrow(OversizeMappingException.class).when(offHeapChainMap).getAndAppend(Matchers.any(), any(ByteBuffer.class));
+
+    when(store.segmentFor(anyLong())).then(new Answer<Object>() {
+      int invocations = 0;
+      @Override
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        if (invocations++ < 10) {
+          return offHeapChainMap;
+        } else {
+          return invocation.callRealMethod();
+        }
+      }
+    });
+    when(store.handleOversizeMappingException(anyLong())).thenReturn(true);
+
+
+    ByteBuffer payload = createPayload(1L);
+
+    store.getAndAppend(1L, payload);
+    assertThat(payload.remaining(), is(8));
+
+    Chain expected = newChainBuilder().build(newElementBuilder().build(payload), newElementBuilder().build(payload));
+    Chain update = newChainBuilder().build(newElementBuilder().build(payload));
+    store.replaceAtHead(1L, expected, update);
+    assertThat(payload.remaining(), is(8));
+  }
+
+  @Test
+  public void test_replaceAtHead_doesNotConsumeBuffer_evenWhenOversizeMappingException() throws Exception {
+    OffHeapServerStore store = (OffHeapServerStore) spy(newStore());
+    final OffHeapChainMap offHeapChainMap = mock(OffHeapChainMap.class);
+    doThrow(OversizeMappingException.class).when(offHeapChainMap).replaceAtHead(Matchers.any(), any(Chain.class), any(Chain.class));
+
+    when(store.segmentFor(anyLong())).then(new Answer<Object>() {
+      int invocations = 0;
+      @Override
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        if (invocations++ < 10) {
+          return offHeapChainMap;
+        } else {
+          return invocation.callRealMethod();
+        }
+      }
+    });
+    when(store.handleOversizeMappingException(anyLong())).thenReturn(true);
+
+
+    ByteBuffer payload = createPayload(1L);
+
+    Chain expected = newChainBuilder().build(newElementBuilder().build(payload), newElementBuilder().build(payload));
+    Chain update = newChainBuilder().build(newElementBuilder().build(payload));
+    store.replaceAtHead(1L, expected, update);
+    assertThat(payload.remaining(), is(8));
   }
 
 }

--- a/clustered/server/src/test/java/org/ehcache/clustered/server/store/ServerStoreTest.java
+++ b/clustered/server/src/test/java/org/ehcache/clustered/server/store/ServerStoreTest.java
@@ -19,6 +19,8 @@ package org.ehcache.clustered.server.store;
 import org.ehcache.clustered.common.internal.store.Chain;
 import org.ehcache.clustered.common.internal.store.Element;
 import org.ehcache.clustered.common.internal.store.ServerStore;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.Is;
 import org.junit.Test;
 
 import java.nio.ByteBuffer;
@@ -51,7 +53,7 @@ public abstract class ServerStoreTest {
     return byteBuffer.getLong();
   }
 
-  private static ByteBuffer createPayload(long key) {
+  protected static ByteBuffer createPayload(long key) {
     ByteBuffer byteBuffer = ByteBuffer.allocate(8).putLong(key);
     byteBuffer.flip();
     return byteBuffer;
@@ -196,6 +198,36 @@ public abstract class ServerStoreTest {
     Chain toVerify = store.get(1);
 
     assertChainAndReverseChainOnlyHave(toVerify, 111, 1111);
+  }
+
+
+  @Test
+  public void test_append_doesNotConsumeBuffer() throws Exception {
+    ServerStore store = newStore();
+    ByteBuffer payload = createPayload(1L);
+
+    store.append(1L, payload);
+    MatcherAssert.assertThat(payload.remaining(), Is.is(8));
+  }
+
+  @Test
+  public void test_getAndAppend_doesNotConsumeBuffer() throws Exception {
+    ServerStore store = newStore();
+    ByteBuffer payload = createPayload(1L);
+
+    store.getAndAppend(1L, payload);
+    MatcherAssert.assertThat(payload.remaining(), Is.is(8));
+  }
+
+  @Test
+  public void test_replaceAtHead_doesNotConsumeBuffer() throws Exception {
+    ServerStore store = newStore();
+    ByteBuffer payload = createPayload(1L);
+
+    Chain expected = newChainBuilder().build(newElementBuilder().build(payload), newElementBuilder().build(payload));
+    Chain update = newChainBuilder().build(newElementBuilder().build(payload));
+    store.replaceAtHead(1L, expected, update);
+    MatcherAssert.assertThat(payload.remaining(), Is.is(8));
   }
 
 }

